### PR TITLE
add allow_manual DAG argument to prevent scheduled DAGs from being triggered manually

### PR DESCRIPTION
**Pull request by houqp**
We have couple very critical DAGs that absolutely cannot be triggered manually ;) 

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

File: airflow/models/dag.py [hunk 1]
```diff
@@ -2347,6 +2352,19 @@ def create_dagrun(
...
+        # prevent triggering manually for scheduled DAGs, while 'allow_manual' = False
+        if (
+            run.run_type == DagRunType.MANUAL
+            and self.schedule_interval is not None
+            and self.schedule_interval != "@once"
+            and not self.allow_manual
+        ):
+            raise AirflowException(
+                f"Cannot manually trigger DAG {self.dag_id}, because it is a scheduled DAG, "
+                f"and 'allow_manual' is set to {self.allow_manual}"
```
Comment (ephraimbuddy) [inline airflow/models/dag.py]: Can we add a unittest so that this is not accidentally removed in the future

**Review by ephraimbuddy — CHANGES_REQUESTED** [review]
This is nice 👍

File: airflow/models/dag.py [hunk 2]
```diff
@@ -2347,6 +2352,19 @@ def create_dagrun(
             creating_job_id=creating_job_id,
             data_interval=data_interval,
         )
+
+        # prevent triggering manually for scheduled DAGs, while 'allow_manual' = False
+        if (
+            run.run_type == DagRunType.MANUAL
+            and self.schedule_interval is not None
+            and self.schedule_interval != "@once"
```
Comment (uranusjr) [inline airflow/models/dag.py]: Easier to use `self.timetable.periodic`

Comment (uranusjr) [reply airflow/models/dag.py]: But currently `allow_manual=False` means “do not allow manual runs _unless the schedule is not periodic_”, which is far from obvious. If this is the semantic you are looking for, the argument should likely be renamed (something like `periodic_only`). If you want to use `allow_manual`, the semantic should be to disable all manual runs without exceptions.

**Comment by tanelk** [comment]
It might be nice to disable the run button in UI, when this option is set.

File: airflow/models/dag.py (hunk 2 above)
Comment (houqp) [reply airflow/models/dag.py]: good call @uranusjr , updated. FYI @XLPI .

File: airflow/models/dag.py (hunk 1 above)
Comment (houqp) [reply airflow/models/dag.py]: I have added a test case 👍

**Review by potiuk — CHANGES_REQUESTED** [review]
Since we are turning that into airflow feature, I think it would be more than reasonable that such DAG have "manual" trigger in the UI disabled. This is far better from the UI point of view than throwing an exception only. It does not follow the "least surprise" principle. Ideally also there should be tooltip on the "DAG Run" button explaining why it is disabled.

I understand this means touching the scary UI code :) which might be dangerous path. Maybe @bbovenzi (or someone from Brent's team) can add a fixup doing it. 

Also this feature needs documentation so that it can actually be discovered by the users: https://airflow.apache.org/docs/apache-airflow/stable/concepts/dags.html

**Comment by ashb** [comment]
Rather than adding another argument to DAG constructor does it maybe make more sense to add this as an argument to the Timetable?

```python
with DAG(dag_id="dag_id", timetable=CronDataIntervalTimetable("0 0 * * *", allow_manual=False):
    ....
```

**Comment by houqp** [comment]
@ashb @uranusjr if we pass the responsibility to `Timetable.infer_manual_data_interval`, how do we enforce this given users can override `infer_manual_data_interval`'s behavior with subclasses?

**Comment by houqp** [comment]
updated code to disable button in the UI

File: airflow/models/dag.py [hunk 3]
```diff
@@ -2280,6 +2284,12 @@ def create_dagrun(
                 "Creating DagRun needs either `run_id` or both `run_type` and `execution_date`"
             )
 
+        if not self.allow_manual and run_type == DagRunType.MANUAL:
+            raise AirflowException(
+                f"Cannot manually trigger DAG {self.dag_id}, because 'allow_manual' is "
+                f"set to {self.allow_manual}"
+            )
```
Comment (uranusjr) [inline airflow/models/dag.py]: Use a dedicated exception subclass here. You will need to catch this in a lot of places, and using `AirflowException` directly will make things very difficult.

File: airflow/exceptions.py [hunk 4]
```diff
@@ -298,3 +298,7 @@ def __repr__(self) -> str:
 
 class TaskDeferralError(AirflowException):
     """Raised when a task failed during deferral for some reason."""
+
+
+class DAGDisallowManual(AirflowException):
```
Comment (uranusjr) [inline airflow/exceptions.py:303]:
```suggestion
class ManualDAGRunDisallowed(AirflowException):
```

seems to read better forme.

**Comment by uranusjr** [comment]
You need to find all `create_dagrun` calls that creates manual runs, and decide how this needs to be catched and handled. I believe there are at least two, one in CLI and one in web UI, and you need to emit the correct response in those situations as appropriate.

**Comment by XLPI** [comment]
@houqp Are we going to follow up regarding this PR further?

**Pull request closed without merging**